contracts/external/cw-token-swap/src/types.rs

```diff
@@ -0,0 +1,420 @@
+            total_amount += amount;
+            Ok(cosmos_msg)
+        })
+        .collect::<Result<Vec<CosmosMsg>, ContractError>>()?;
+
+    // Verify that total amount of funds matches funds sent in all messages
+    if total_amount != amount {
+        return Err(ContractError::WrongFundsCalculation {});
+    }
+    Ok(cosmos_msgs)
+}
+
+#[cw_serde]
+pub enum NativeSendMsg {
+    BankSend {
+        to_address: String,
+        amount: Vec<Coin>,
+    },
+    BankBurn {
+        amount: Vec<Coin>,
+    },
+    WasmExecute {
+        contract_addr: String,
+        msg: Binary,
+        funds: Vec<Coin>,
+    },
+    WasmInstantiate {
+        admin: Option<String>,
+        code_id: u64,
+        msg: Binary,
+        funds: Vec<Coin>,
+        label: String,
+    },
+}
+
+impl CompletionMsgsToCosmosMsg for NativeSendMsg {
+    fn into_checked_cosmos_msg(
+        self,
+        deps: Deps,
+        denom: &str,
+    ) -> Result<(Uint128, CosmosMsg), ContractError> {
+        let verify_coin = |coins: &Vec<Coin>| {
+            if coins.len() != 1 {
+                return Err(ContractError::InvalidSendMsg {});
+            }
+            if coins[0].amount.is_zero() {
+                return Err(ContractError::InvalidSendMsg {});
+            }
+            if denom != coins[0].denom {
+                return Err(ContractError::InvalidSendMsg {});
+            }
+            Ok(coins[0].amount)
+        };
+
+        match self {
+            NativeSendMsg::BankSend { to_address, amount } => Ok((
+                verify_coin(&amount)?,
+                BankMsg::Send {
+                    to_address: deps.api.addr_validate(&to_address)?.to_string(),
+                    amount,
+                }
+                .into(),
+            )),
+            NativeSendMsg::BankBurn { amount } => {
+                Ok((verify_coin(&amount)?, BankMsg::Burn { amount }.into()))
+            }
+            NativeSendMsg::WasmExecute {
+                contract_addr,
+                msg,
+                funds,
+            } => Ok((
+                verify_coin(&funds)?,
+                WasmMsg::Execute {
+                    contract_addr: deps.api.addr_validate(&contract_addr)?.to_string(),
+                    msg,
+                    funds,
+                }
+                .into(),
+            )),
+            NativeSendMsg::WasmInstantiate {
+                admin,
+                code_id,
+                msg,
+                funds,
+                label,
+            } => Ok((
+                verify_coin(&funds)?,
+                WasmMsg::Instantiate {
+                    admin: admin.map(|a| deps.api.addr_validate(&a).unwrap().to_string()),
+                    code_id,
+                    msg,
+                    funds,
+                    label,
+                }
+                .into(),
+            )),
+        }
+    }
+}
+
+#[cw_serde]
+pub enum Cw20SendMsgs {
```

Comment:
i'm curious why you chose to break these messages into multiple enums and use an interface, instead of just using one enum? for example:

```
pub enum SomeMessages {
  Cw20(Cw20SendMsgs),
  Native(NativeSendMsgs),
}
```

personally, I am pretty anti traits and OOP style code as I think its less readable and thus more dangerous.

Reply:
Actually this was done to avoid doing the same checks 2 times and nothing more, I understand it totally added unnecessary confusion/complexity to it, I will remove it.